design/20230719-cabundle-reference.md

```diff
@@ -0,0 +1,120 @@
+---
+title: Certificate Request CRD
+authors:
+  - "@AppalaKarthik"
+  - "@sankalp-at-github"
+creation-date: 2023-07-19
+last-updated: 2023-07-19
+#status: implementable
+---
+
+# Design: Certficate Authority Bundle Reference Using Secrets, Configmaps
+
+
+<!-- toc -->
+- [Problem Statement](#problem-statement)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Graduation Criteria](#graduation-criteria)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Problem Statement
+
+In Issuer/ClusterIssuer resources caBundle field must be specified as a base64 encoded string.
+
+```
+spec:
+  venafi:
+    tpp:
+      caBundle: <B64_ENCODED_STRING>
+      credentialsRef:
+        name: tpp-token
+      url: https://my-server.com/vedsdk/
+    zone: Certificates\public
+```
+
+This becomes a configuration management issue when many issuers are there in a cluster using the same TPP endpoint, the same CA information has to be copied in all the issuers and there can be mistakes made as this would be manual.
+
+## Summary
+
+Ability to reference caBundle from the contents of secrets, configmap.
+
+## Motivation
+The proposed change would encourage the way caBundle is referenced in kubernetes native way.
+
+### Goals
+
+To be able to refer caBundle using secret.
+To be able to refer caBundle using configmap.
+
+### Non-Goals
+
+To include reference using resources other than mentioned.
+
+## Proposal
+
+Enable reading of cabundle from native kubernetes objects like secrets and configmaps.
+
+## Design Details
+
+Add secrets and configMaps as optional fields to the spec which can be used to refer a key to find the caBundle. Below is the example for venafi issuer.
+
+```
+spec:
+  venafi:
+    tpp:
+      caBundle: <B64_ENCODED_STRING>
+      caBundleSecretRef:
+        name: <>
+        key: <>
+      caBundleConfigMapRef:
+        name: <>
+        key: <>
+      credentialsRef:
+        name: tpp-token
+      url: https://my-server.com/vedsdk/
+    zone: Certificates\public
+```
+
+
+### Graduation Criteria
+
+Placing the `caBundleSecetRef` and `caBundleConfigMapRef` specification functionality behind a feature gate should be required.
```

Comment:
I'm not sure that we have to introduce a new feature flag for this.
We also did not introduce a feature flag when we added secret refs for Vault: https://github.com/cert-manager/cert-manager/pull/5387

Reply:
Are we watching configmaps in cert-manager already? If not, I think it could make sense to initially add it behind a feature gate.